wallet/src/repository/wallet/mod.rs

```diff
@@ -785,6 +877,171 @@ fn convert_block_epoch_to_timestamp(epoch_constants: EpochConstants, epoch: Epoc
+                })
+                .collect::<Vec<model::Output>>(),
+        }),
+        types::Transaction::Mint(mint) => model::TransactionData::Mint(model::MintData {
+            output: model::Output {
+                address: mint.output.pkh.to_string(),
+                time_lock: mint.output.time_lock,
+                value: mint.output.value,
+            },
+        }),
+        types::Transaction::Tally(tally) => model::TransactionData::Tally(model::TallyData {
+            request_transaction_hash: tally.dr_pointer.to_string(),
+            outputs: tally
+                .outputs
+                .clone()
+                .into_iter()
+                .map(|output| model::Output {
+                    address: output.pkh.to_string(),
+                    time_lock: output.time_lock,
+                    value: output.value,
+                })
+                .collect::<Vec<model::Output>>(),
+            tally: build_tally_report(tally, &txn.metadata)?,
+        }),
+        _ => {
+            return Err(Error::UnsupportedTransactionType(format!(
+                "{:?}",
+                txn.transaction
+            )));
+        }
+    };
+
+    Ok(model::BalanceMovement {
+        kind,
+        amount,
+        transaction: model::Transaction {
+            hash: hex::encode(txn.transaction.hash()),
+            timestamp,
+            block: Some(model::Beacon {
+                epoch: block_info.epoch,
+                block_hash: block_info.block_hash,
+            }),
+            miner_fee,
+            data: transaction_data,
+        },
+    })
+}
+
+fn build_tally_report(
+    tally: &types::TallyTransaction,
+    metadata: &Option<model::TransactionMetadata>,
+) -> Result<model::TallyReport> {
+    let reveals = match metadata {
+        Some(model::TransactionMetadata::Tally(report)) => {
+            // List of reveals extracted from Data Request Report
+            let mut reveals: HashMap<PublicKeyHash, model::Reveal> = report
+                .reveals
+                .iter()
+                .map(|(pkh, reveal_txn)| {
+                    types::RadonTypes::try_from(reveal_txn.body.reveal.as_slice())
+                        .map(|x| {
+                            (
+                                *pkh,
+                                model::Reveal {
+                                    value: x.to_string(),
+                                    in_consensus: true,
+                                },
+                            )
+                        })
+                        .map_err(|err| Error::RevealRadDecode(err.to_string()))
+                })
+                .collect::<Result<HashMap<PublicKeyHash, model::Reveal>>>()?;
+
+            // Set not in consensus reveals
+            for pkh in &tally.slashed_witnesses {
+                let liar = reveals.get(&pkh).cloned();
+                if let Some(reveal) = liar {
+                    reveals.insert(
```

Comment:
It looks like you are trying to mutate. You can use `reveals.get_mut()` and then `reveal.value.in_consensus = false`.

Reply:
And in the case where someone is penalized but has no reveal, you can try to insert a `RadonTypes::RadonError(NoReveal)`.

Reply:
I like it!